### PR TITLE
[WIP] Add remote file io using fsspec.

### LINKED ISSUE #33452: ENH: Use fsspec for reading/writing from/to S3, GCS, Azure Blob, etc.
#### Is your feature request related to a problem?

Currently pandas has some support for S3 and GCS using the pandas.io.{gcs,s3} modules, which are based on S3fs and gcsfs. 

It seems like we could easily broaden the support for different filesystems by leveraging the fsspec library (https://pypi.org/project/fsspec/) and its interface implementations (see https://github.com/intake/filesystem_spec/blob/master/fsspec/registry.py for some examples) to read/write files in pandas. 

This way, we would also be able to use filesystems such as Azure-based storage systems directly from pandas. 

#### Describe the solution you'd like

I'd like to be able to use the different file systems supported by fsspec in pandas with something like:

```
import pandas as pd

df1 = pd.read_csv("abfs://my_container/my_file.csv")
df1.to_json("file:///some/local/path.json") # Also works without file:// prefix.

df2 = pd.read_csv("s3://my_bucket/my_file.csv")
...
```

#### API breaking implications

In principle, it looks as if we could cover most of the work by adapting get_filepath_or_buffer in pandas/io/common.py to use fsspec. We would of course have to test if fsspec doesn't break anything compared to the current implementations.

One challenge is that some storage systems require extra arguments (called storage options in fsspec). For example, Azure blob requires the user to pass two storage options (account_name and account_key) to be able to access the storage. We would need to consider how to pass these options to the correct methods, either by (a) setting these options globally for a given type of storage or (b) passing the options through the pd.read_* functions and pd.DataFrame.to_* methods. 

#### Describe alternatives you've considered

This seems like a change that would have a small impact, as pandas already uses S3fs and gcsfs, which are both implementations of the broader fsspec interface. It should also provide support for a great number of different filesystems with minimal changes, compared to adding support for each filesystem ourselves. As such, I would think it is the preferred approach. 

Another approach would be to add support for each additional filesystem as it comes along. This however would require adding and maintaining code for each filesystem, which seems less preferable.

#### Additional context

I recently implemented similar wrappers for pandas code at one of our clients, and am therefore somewhat familiar with fsspec etc. I would be happy to see if we can contribute these ideas + code to the pandas project.

### PR DESCRIPTION
**Pull request by jrderuiter**
- [x] closes #33452
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/io/common.py [hunk 1]
```diff
@@ -194,19 +211,26 @@ def get_filepath_or_buffer(
         req.close()
         return reader, encoding, compression, True
 
-    if is_s3_url(filepath_or_buffer):
-        from pandas.io import s3
+    if is_fsspec_url(filepath_or_buffer):
+        import fsspec
+        scheme = parse_url(filepath_or_buffer).scheme
+        filesystem = fsspec.filesystem(scheme)
+        file_obj = filesystem.open(filepath_or_buffer, mode=mode or "rb")
```
Comment (alimcmaster1) [inline pandas/io/common.py:218]: Do we need to pass through the encoding? to .open() You will potentially also fix this if you do so: https://github.com/pandas-dev/pandas/issues/26124 :)

**Review by TomAugspurger — COMMENTED** [review]
Have you been able to test this out manually? Do things seem OK?

**Comment by martindurant** [comment]
@jrderuiter , is there any way in which I can help here?

**Comment by martindurant** [comment]
ping ?

File: pandas/io/common.py [hunk 2]
```diff
@@ -158,6 +158,23 @@ def urlopen(*args, **kwargs):
     return urllib.request.urlopen(*args, **kwargs)
 
 
+def is_fsspec_url(url) -> bool:
```
Comment (jreback) [inline pandas/io/common.py:161]: can u type url

File: pandas/io/common.py [hunk 3]
```diff
@@ -158,6 +158,23 @@ def urlopen(*args, **kwargs):
...
+    """
+    Returns true if fsspec is installed and the URL references a known
+    fsspec filesystem.
+    """
+
+    if not isinstance(url, str):
```
Comment (jreback) [inline pandas/io/common.py:167]: you don’t need this

**Review by jreback — CHANGES_REQUESTED** [review]
you need to add fsspec to several of the ci files but not all and environment.yaml

to have testing and proper skipping

**Comment by TomAugspurger** [comment]
Closing in favor of https://github.com/pandas-dev/pandas/pull/34266.

**Pull request closed without merging**